hs_core/models.py

```diff
@@ -1508,41 +1512,205 @@ class Meta:
+                relpath.startswith(self.content_object.resource_federation_path + '/'):
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(self.content_object.resource_federation_path)
+            relpath = relpath[plen+1:]  # omit /
+
+            # strip resource id from path
+            if relpath.startswith(file_path):
+                plen = len(file_path)
+                relpath = relpath[plen:]  # omit /
+            else:
+                raise ValidationError("Malformed federated resource path")
+        elif path.startswith(file_path):
+            # strip optional local path prefix
+            if not storage.exists(path):
+                raise ValidationError("Federated path does not exist in irods")
+            plen = len(file_path)
+            relpath = relpath[plen:]  # strip local prefix, omit /
+
+        # now we have folder/file. We could have gotten this from the input, or
+        # from stripping qualification folders. Note that this can contain
+        # misnamed header content misinterpreted as a folder
+        if '/' in relpath:
+            folder, base = relpath.rsplit('/', 1)
+            abspath = get_path(self, base, folder=folder)
+            print("abspath = " + abspath)
+            if not storage.exists(abspath):
+                raise ValidationError("Local path does not exist in irods")
+        else:
+            folder = None
+            base = relpath
+            abspath = get_path(self, base, folder=folder)
+            print("abspath = " + abspath)
+            if not storage.exists(abspath):
+                raise ValidationError("Local path does not exist in irods")
+
+        return folder, base
+
+    # def rename(self, new_name):
+    #     """ rename a file, setting all path variables appropriately """
+    #     pass
+
+    # def copy_irods(self, source_path, dest_path=None):
+    #     """ copy an irods file into this FileField, setting all paths appropriately """
+    #     pass
+
+    # def move_irods(self, source_path, dest_path=None):
+    #     """ move an irods file into this object, setting all paths appropriately """
+
+    # classmethods do things that affect all files.
+
+    @classmethod
+    def create_folder(cls, resource, folder):
+        """ create a folder for a resource """
+        # avoid import loop
+        from hs_core.views.utils import create_folder
+        _path_is_allowed(folder)
+        # TODO: move code from location used below to here
+        create_folder(resource.short_id, folder)
+
+    @classmethod
+    def remove_folder(cls, resource, folder, user):
+        """ remove a folder for a resource """
+        # avoid import loop
+        from hs_core.views.utils import remove_folder
+        _path_is_allowed(folder)
+        # TODO: move code from location used below to here
+        remove_folder(user, resource.short_id, folder)
```

Comment:
@alvacouch Although it is always nice to isolate iRODS interaction to one class, I don't think putting these folder related operations in ResourceFile class is appropriate since semantically it is not correct as ResourceFile as currently defined only corresponds to one single resource file object and these folder operations are beyond one single file but multiple files and folders under a resource. So I think BaseResource class is a better place to put them in. Or make it better encapsulated, perhaps add a ResourceFiles OneToOneField to BaseResource class and make each ResourceFile a ForeignKey many-to-one relationship to ResourceFiles. However, I am afraid we would have troubles in data migration if we make a change like that.